text/0032-dust-protection/0032-dust-protection.md

```diff
@@ -0,0 +1,114 @@
+                                                <table>
+                                                    <tr>
+                                                        <td><b>Name</b></td>
+                                                        <td><b>Type</b></td>
+                                                        <td><b>Description</b></td>
+                                                    </tr>
+                                                    <tr>
+                                                        <td>Address Type</td>
+                                                        <td>uint8</td>
+                                                        <td>
+                                                            Set to <strong>value 1</strong> to denote an <i>Ed25519 Address</i>.
+                                                        </td>
+                                                    </tr>
+                                                    <tr>
+                                                        <td>Address</td>
+                                                        <td>Array&lt;byte&gt;[32]</td>
+                                                        <td>The raw bytes of the Ed25519 address which is a BLAKE2b-256 hash of the Ed25519 public key.</td>
+                                                    </tr>
+                                                </table>
+                                            </details>
+                                        </td>
+                                    </tr>
+                                    <tr>
+                                        <td>Amount</td>
+                                        <td>uint64</td>
+                                        <td>The amount of tokens to deposit with this <i>SigLockedSingleOutput</i> output.</td>
+                                    </tr>
+                                </table>
+
+
+### The new protocol dust rule
+
+Let `A` be the address that should hold the dust outputs' balances. Let `M` be the sum of all SigLockedDustAllowanceOutputs' values on `A`.So the total allowed dust outputs on `A` is (`M`/1 MIOTA) * 100 rounded down to the nearest integer. So 100 outputs for 1 MIOTA deposited.
+
+The SigLockedDustAllowanceOutput must be greater or equal to 1MI.
+
+If while processing a milestone a new dust UTXO is created but there is not a sufficient amount already locked in SigLockedDustAllowanceOutputs then the encompassing message should be marked as `ignored`.
+
+If while processing a milestone an unspent SigLockedDustAllowanceOutput is consumed and as a result we have a violation on the number of allowed dust outputs then the encompassing message should be marked as `ignored`.
+
+The checks should happen after the entire message was processed. Messages that were previously applied by the current milestone are taken into account.
```

Comment:
```suggestion
Let A be the address that should hold the dust outputs' balances. Let S be the sum of all the amounts of all unspent `SigLockedDustAllowanceOutputs` on A. Then, the maximum number of allowed dust outputs on A is S divided by 10,000 and rounded down, i.e. 100 outputs for each 1 Mi deposited.

The amount of a `SigLockedDustAllowanceOutput` must be at least 1 Mi. Apart from this, `SigLockedDustAllowanceOutputs` are processed identical to `SigLockedOutputs`. The transaction validation as defined in [Draft RFC-18](https://github.com/luca-moser/protocol-rfcs/blob/signed-tx-payload/text/0000-transaction-payload/0000-transaction-payload.md), however, needs to be adapted.

_Syntactical validation_ for `SigLockedDustAllowanceOutput`:
- The `Address` must be unique in the set of `SigLockedSingleOutputs` and `SigLockedDustAllowanceOutput`.
- The `Amount` must be ≥ 1,000,000.

The _semantic validation_ remains unchanged and are checked for both `SigLockedSingleOutputs` and `SigLockedDustAllowanceOutput`, but this RFC introduces one additional criterion:

A transaction T
  - consuming a `SigLockedDustAllowanceOutput` on address A **or**
  - creating a dust output with address A,

is only semantically valid, if, after T is booked, the number of unspent dust outputs on A must not exceed the allowed threshold of S / 10,000. When T contains more than one such output, the threshold must not be exceeded on any corresponding address.
```
Made the validation rules more technical and consistent with the way it is described in the transaction RFC

Reply:
> Apart from this, `SigLockedDustAllowanceOutputs` are processed identical to `SigLockedDustAllowanceOutput`. 

?

Reply:
> - The `Address` must be unique in the set of `SigLockedSingleOutputs` and `SigLockedDustAllowanceOutput`.

This does not need to be true as long as each output of type `SigLockedDustAllowanceOutput` has an amount of >= 1,000,000 .

Reply:
is only semantically valid, if, after T is booked, the number of unspent dust outputs on A ~~must~~ **does** not exceed the allowed threshold of S / 10,000.

---

When T contains more than one such ~~output~~ input / consumes more than one such output

Reply:
> You mean `TransactionPayload`?

What I mean is that in #18 any mention of `SigLockedSingleOutput` is supposed to be replaced with "`SigLockedSingleOutput` and/or `SigLockedDustAllowanceOutput`". But I didn't wanna go through all occurrences explicitly. Any Ideas how to makes this more clear.

Reply:
> This does not need to be true as long as each output of type `SigLockedDustAllowanceOutput` has an amount of >= 1,000,000 .

Currently #18 forbids creating a transaction with more than one output to the same address. We should try to preserve this somehow. But indeed since we now have more than one output type the question is, do we want to allow a transaction with 1 `SigLockedDustAllowanceOutput` to A and 1 `SigLockedSingleOutput` to A? Or is this forbidden?

Definitely we should forbid more than one  `SigLockedDustAllowanceOutput` to A in one transaction, just we currently do for `SigLockedSingleOutput`.

Reply:
> When T contains more than one such ~output~ input / consumes more than one such output

Not 100% sure whether I understand this remark. I made some minimal changes to the previously suggested changes, is that better now?

Reply:
Not sure I understand that last sentence you added...
at least I can interpret it more than 1 way

Reply:
OK then. I've removed that sentence, it is redundant anyways, I was just trying to make it more explicit that when a transaction contains multiple dust outputs on different addresses, the criteria must hold for all of them.